File: nodejs/1352_productOfNumbers.ts
// https://leetcode.com/problems/product-of-the-last-k-numbers/
// 14/02/2025

class ProductOfNumbers {
    products: number[];

    constructor() {
        this.products = [];
    }

    add(num: number): void {
        if (num === 0) {
            this.products.length = 0; // clear the array
        } else if (this.products.length === 0) {
            this.products.push(num);
        } else {
            const p = this.products[this.products.length - 1];
            this.products.push(num * p);
        }
    }

    getProduct(k: number): number {
        const n = this.products.length;
        if (n === 0 || k > n) {
            return 0;
        }

        if (n === k) {
            return this.products[n - 1];
        }

        return this.products[n - 1] / this.products[n - k - 1];
    }
}


// (() => {
//     const testCases: [number, number[][], number[]][] = [
//         [
//             4,
//             [[1, 4], [2, 5], [1, 3], [3, 4]],
//             [1, 2, 2, 3]
//         ],
//         [
//             4,
//             [[0, 1], [1, 2], [2, 2], [3, 4], [4, 5]],
//             [1, 2, 2, 3, 4]
//         ],
//     ];

//     testCases.forEach(([limit, queries, expected], i) => {
//         console.log(`Test Case #${i}:`);
//         const result = queryResults(limit, queries);
//         console.log(`Input: ${[limit, queries]}, Expected: ${expected}, Got: ${result}`);
//         _.isEqual(result, expected) ? console.log("PASS") : console.log("FAIL");
//     })
// })();

File: rust/src/bin/1352_product_of_numbers.rs
// https://leetcode.com/problems/product-of-the-last-k-numbers/
// 14/02/2025

#[derive(Debug, Clone)]
struct ProductOfNumbers {
    products: Vec<i32>,
}

impl ProductOfNumbers {
    fn new() -> Self {
        ProductOfNumbers {
            products: Vec::new(),
        }
    }

    fn add(&mut self, num: i32) {
        match num {
            0 => self.products.clear(),
            _ => match self.products.last() {
                Some(p) => self.products.push(p * num),
                None => self.products.push(num),
            },
        }
    }

    fn get_product(&self, k: i32) -> i32 {
        let (n, k) = (self.products.len(), k as usize);
        if n == 0 || k > n {
            return 0;
        }

        if n == k {
            return self.products[n - 1];
        }

        return self.products[n - 1] / self.products[n - k - 1];
    }
}

fn main() {
    let test_cases = vec![(
        vec![
            "ProductOfNumbers",
            "add",
            "add",
            "add",
            "add",
            "add",
            "getProduct",
            "getProduct",
            "getProduct",
            "add",
            "getProduct",
        ],
        vec![
            vec![],
            vec![3],
            vec![0],
            vec![2],
            vec![5],
            vec![4],
            vec![2],
            vec![3],
            vec![4],
            vec![8],
            vec![2],
        ],
        vec![
            None,
            None,
            None,
            None,
            None,
            None,
            Some(20),
            Some(40),
            Some(0),
            None,
            Some(32),
        ],
    )];

    for (instruction, param, expected) in test_cases {
        let result: Vec<Option<i32>> = instruction
            .iter()
            .zip(param.iter())
            .scan(ProductOfNumbers::new(), |pn, (i, p)| match i {
                &"add" => {
                    pn.add(p[0]);
                    Some(None)
                }
                &"getProduct" => Some(Some(pn.get_product(p[0]))),
                _ => Some(None),
            })
            .collect();
        println!(
            "Input: {:?}, Expected: {:?}, Got: {:?}",
            (instruction, param),
            expected,
            result
        );
        assert_eq!(result, expected);
    }
}
